### DISCUSSION TITLE
Add red asterisk to TextInput required field

### DISCUSSION THREAD

**Discussion by meliodev** [callstack/react-native-paper#2591] [Q&A] [answered]
Hey guys! Is there any way to add asterisk with red color to textInput label (placeholder) to show required field ?

**Comment by raajnadar** [comment, marked as answer]
NEW UPDATE -

There is a new way to do it from  paper 4.11.0 - https://github.com/callstack/react-native-paper/pull/2991

We can use custom react element for the text input component. Thanks to @smbadiwe


FOR OLD VERSIONS -

```
 <TextInput
      label="Email *"
      value={text}
      onChangeText={text => setText(text)}
 />
```

**Reply by irshadpalayadan** [reply]
ideally there should be `required` props as part of the TextInput

**Reply by raajnadar** [reply]
@irshadpalayadan answer updated

**Comment by smbadiwe** [comment]
The accepted answer is not really a solution. The requirement is to have a styled (colored) asterisk, not just an asterisk.

The quickest solution is to update this codebase to change the `label` property type from `string | undefined` to `React.ReactNode | undefined`. This will allow developers to customize how they want to indicate to their app users that a field is required.

In plain Javascript, the following example will work (you can try it by modifying [this Expo example](https://snack.expo.dev/?name=TextInput&description=https%3A%2F%2Fcallstack.github.io%2Freact-native-paper%2Ftext-input.html&code=import%20*%20as%20React%20from%20%27react%27%3B%0Aimport%20%7B%20TextInput%20%7D%20from%20%27react-native-paper%27%3B%0A%0Aconst%20MyComponent%20%3D%20()%20%3D%3E%20%7B%0A%20%20const%20%5Btext%2C%20setText%5D%20%3D%20React.useState(%27%27)%3B%0A%0A%20%20return%20(%0A%20%20%20%20%3CTextInput%0A%20%20%20%20%20%20label%3D%22Email%22%0A%20%20%20%20%20%20value%3D%7Btext%7D%0A%20%20%20%20%20%20onChangeText%3D%7Btext%20%3D%3E%20setText(text)%7D%0A%20%20%20%20%2F%3E%0A%20%20)%3B%0A%7D%3B%0A%0Aexport%20default%20MyComponent%3B&dependencies=react-native-paper):
```
<TextInput
      label={<><span style={{color:'red'}}>*</span> Email</>}
      value={text}
      onChangeText={text => setText(text)}
    />
```

But it will fail in typescript due to type-checking.

**Reply by raajnadar** [reply]
You're typescript type is wrong that is why the type check will fail.

**Reply by smbadiwe** [reply]
> You're typescript type is wrong that is why the type check will fail.

Yes, and that's the problem. `label` prop is defined in the library as of datatype `string` ([See Source Code](https://github.com/callstack/react-native-paper/blob/b545cdcbd8c5f1bd5ad3f0e9f095c294527846a4/src/components/TextInput/TextInput.tsx#L41)), which limits customization. If the datatype of `label` were to be changed to `React.ReactNode` (which also supports plain strings), then one could pass in any custom component and that'll perfectly solve this problem.

**Reply by raajnadar** [reply]
The repo is open for PR, you can fix it

**Reply by smbadiwe** [reply]
@raajnadar PR is up at https://github.com/callstack/react-native-paper/pull/2991

**Reply by raajnadar** [reply]
Thanks for the PR & getting that merged I updated my previous comment.
